File: indicators.py
"""
Technical indicators, analytics, and market regime classification implementation.
Operates completely autonomously with dynamic parameter adaptation and optional
Rust CFFI acceleration for maximum institutional execution performance.
"""

import math


def calculate_ema(prices, period):
    """
    Calculates Exponential Moving Average (EMA) with Rust CFFI acceleration when available.
    Formula: EMA_today = Price_today * multiplier + EMA_yesterday * (1 - multiplier)
    multiplier = 2 / (period + 1)
    """
    if not prices or period <= 0 or len(prices) < period:
        return None

    try:
        from institutional_integrations.rust_bridge import is_rust_available, rust_accelerated_ema
        if is_rust_available():
            ema_series = rust_accelerated_ema(prices, period)
            if ema_series and len(ema_series) == len(prices):
                return ema_series[-1]
    except Exception:
        pass

    multiplier = 2.0 / (period + 1)
    # Start with simple moving average as first EMA value
    sma = sum(prices[:period]) / float(period)
    ema = sma

    for price in prices[period:]:
        ema = (price * multiplier) + (ema * (1.0 - multiplier))

    return ema


def calculate_rsi(prices, period=14):
    """
    Calculates Relative Strength Index (RSI).
    """
    if not prices or len(prices) < period + 1 or period <= 0:
        return 50.0

    gains = []
    losses = []

    for i in range(1, len(prices)):
        diff = prices[i] - prices[i - 1]
        if diff > 0:
            gains.append(diff)
            losses.append(0.0)
        else:
            gains.append(0.0)
            losses.append(abs(diff))

    if len(gains) < period:
        return 50.0

    avg_gain = sum(gains[:period]) / float(period)
    avg_loss = sum(losses[:period]) / float(period)

    if avg_loss == 0.0:
        return 100.0 if avg_gain > 0 else 50.0

    for i in range(period, len(gains)):
        avg_gain = (avg_gain * (period - 1) + gains[i]) / float(period)
        avg_loss = (avg_loss * (period - 1) + losses[i]) / float(period)

    if avg_loss == 0.0:
        return 100.0

    rs = avg_gain / avg_loss
    rsi = 100.0 - (100.0 / (1.0 + rs))
    return round(rsi, 4)


def calculate_atr(highs, lows, closes, period=14):
    """
    Calculates Average True Range (ATR) with Wilder's smoothing.
    """
    if not closes or len(closes) < period + 1 or period <= 0:
        return None

    true_ranges = []
    for i in range(1, len(closes)):
        h = highs[i]
        l = lows[i]
        prev_c = closes[i - 1]

        tr1 = h - l
        tr2 = abs(h - prev_c)
        tr3 = abs(l - prev_c)

        true_ranges.append(max(tr1, tr2, tr3))

    if len(true_ranges) < period:
        return None

    atr = sum(true_ranges[:period]) / float(period)
    for i in range(period, len(true_ranges)):
        atr = (atr * (period - 1) + true_ranges[i]) / float(period)

    return atr


def calculate_macd(prices, fast_period=12, slow_period=26, signal_period=9):
    """
    Calculates MACD (Moving Average Convergence Divergence).
    Returns dict: { 'macd': float, 'signal': float, 'histogram': float } or None.
    """
    if not prices or len(prices) < slow_period + signal_period:
        return None

    fast_emas = []
    slow_emas = []

    for i in range(slow_period, len(prices) + 1):
        window = prices[:i]
        fast_ema = calculate_ema(window, fast_period)
        slow_ema = calculate_ema(window, slow_period)
        if fast_ema is not None and slow_ema is not None:
            fast_emas.append(fast_ema)
            slow_emas.append(slow_ema)

    macd_line = [f - s for f, s in zip(fast_emas, slow_emas)]

    if len(macd_line) < signal_period:
        return None

    signal_line = calculate_ema(macd_line, signal_period)
    if signal_line is None:
        return None

    current_macd = macd_line[-1]
    current_signal = signal_line
    current_histogram = current_macd - current_signal

    return {
        "macd": current_macd,
        "signal": current_signal,
        "histogram": current_histogram,
    }


def calculate_bollinger_bands(prices, period=20, num_std=2.0):
    """
    Calculates Bollinger Bands.
    Returns dict: { 'upper': float, 'middle': float, 'lower': float, 'bandwidth': float } or None.
    """
    if not prices or len(prices) < period or period <= 0:
        return None

    window = prices[-period:]
    middle = sum(window) / float(period)

    variance = sum((x - middle) ** 2 for x in window) / float(period)
    std_dev = math.sqrt(max(0.0, variance))
    if std_dev == 0.0:
        std_dev = 1e-9

    upper = middle + (num_std * std_dev)
    lower = middle - (num_std * std_dev)
    bandwidth = (upper - lower) / middle if middle != 0 else 0.0

    return {
        "upper": upper,
        "middle": middle,
        "lower": lower,
        "bandwidth": bandwidth,
    }


def calculate_pivot_points(high, low, close):
    """
    Calculates classic floor support and resistance pivot points.
    Returns dict: { 'pivot': float, 'r1': float, 's1': float, 'r2': float, 's2': float }
    """
    pivot = (high + low + close) / 3.0
    r1 = (2.0 * pivot) - low
    s1 = (2.0 * pivot) - high
    r2 = pivot + (high - low)
    s2 = pivot - (high - low)
    return {"pivot": pivot, "r1": r1, "s1": s1, "r2": r2, "s2": s2}


def calculate_donchian_channels(highs, lows, period=20):
    """
    Calculates Donchian Channels (highest high and lowest low over lookback).
    Returns dict: { 'upper': float, 'lower': float, 'middle': float } or None.
    """
    if not highs or not lows or len(highs) < period or len(lows) < period or period <= 0:
        return None

    upper = max(highs[-period:])
    lower = min(lows[-period:])
    middle = (upper + lower) / 2.0
    return {"upper": upper, "lower": lower, "middle": middle}


def calculate_bollinger_squeeze(prices, period=20, num_std=2.0):
    """
    Calculates Bollinger Band Width ratio to evaluate volatility squeeze conditions.
    Formula: (Upper Band - Lower Band) / Middle Band
    Returns float or None.
    """
    bb = calculate_bollinger_bands(prices, period, num_std)
    if bb is None or bb["middle"] == 0:
        return None
    return bb["bandwidth"]


def calculate_adx(highs, lows, closes, period=14):
    """
    Average Directional Index (ADX) measuring trend strength.
    Range [0, 100]. Values > 22-25 indicate strong trend.
    """
    if not closes or len(closes) < period * 2 or len(highs) < period * 2 or len(lows) < period * 2:
        return 20.0

    tr_list = []
    dm_plus = []
    dm_minus = []

    for i in range(1, len(closes)):
        h_diff = highs[i] - highs[i - 1]
        l_diff = lows[i - 1] - lows[i]

        tr = max(
            highs[i] - lows[i],
            abs(highs[i] - closes[i - 1]),
            abs(lows[i] - closes[i - 1]),
        )
        tr_list.append(tr)

        if h_diff > l_diff and h_diff > 0:
            dm_plus.append(h_diff)
        else:
            dm_plus.append(0.0)

        if l_diff > h_diff and l_diff > 0:
            dm_minus.append(l_diff)
        else:
            dm_minus.append(0.0)

    if len(tr_list) < period:
        return 20.0

    tr_smooth = [sum(tr_list[:period])]
    dm_plus_smooth = [sum(dm_plus[:period])]
    dm_minus_smooth = [sum(dm_minus[:period])]

    for i in range(period, len(tr_list)):
        tr_smooth.append(tr_smooth[-1] - (tr_smooth[-1] / float(period)) + tr_list[i])
        dm_plus_smooth.append(
            dm_plus_smooth[-1] - (dm_plus_smooth[-1] / float(period)) + dm_plus[i]
        )
        dm_minus_smooth.append(
            dm_minus_smooth[-1] - (dm_minus_smooth[-1] / float(period)) + dm_minus[i]
        )

    dx_list = []
    for i in range(len(tr_smooth)):
        tr_val = tr_smooth[i] if tr_smooth[i] > 0 else 1e-5
        di_p = (dm_plus_smooth[i] / tr_val) * 100.0
        di_m = (dm_minus_smooth[i] / tr_val) * 100.0

        diff = abs(di_p - di_m)
        total = di_p + di_m if (di_p + di_m) > 0 else 1e-5
        dx_list.append((diff / total) * 100.0)

    if len(dx_list) < period:
        return 20.0

    adx = sum(dx_list[:period]) / float(period)
    for i in range(period, len(dx_list)):
        adx = ((adx * (period - 1)) + dx_list[i]) / float(period)

    return round(adx, 2)


def calculate_stochastic(highs, lows, closes, period=14, d_period=3):
    """
    Stochastic Oscillator (%K and %D).
    Returns dict: {'k': float, 'd': float}
    """
    if not closes or len(closes) < period:
        return {"k": 50.0, "d": 50.0}

    k_values = []
    for i in range(len(closes) - d_period, len(closes)):
        sub_highs = highs[max(0, i - period + 1) : i + 1]
        sub_lows = lows[max(0, i - period + 1) : i + 1]
        if not sub_highs or not sub_lows:
            continue
        h_high = max(sub_highs)
        l_low = min(sub_lows)
        denom = h_high - l_low
        k = 50.0 if denom == 0 else ((closes[i] - l_low) / float(denom)) * 100.0
        k_values.append(k)

    k_curr = k_values[-1] if k_values else 50.0
    d_curr = sum(k_values) / float(len(k_values)) if k_values else 50.0
    return {"k": round(k_curr, 2), "d": round(d_curr, 2)}


def calculate_ichimoku(highs, lows, closes):
    """
    Ichimoku Cloud key parameters (Tenkan-sen and Kijun-sen).
    Returns dict: {'tenkan': float, 'kijun': float}
    """
    if not closes or len(closes) < 26:
        curr = closes[-1] if closes else 0.0
        return {"tenkan": curr, "kijun": curr}

    high_9 = max(highs[-9:])
    low_9 = min(lows[-9:])
    tenkan = (high_9 + low_9) / 2.0

    high_26 = max(highs[-26:])
    low_26 = min(lows[-26:])
    kijun = (high_26 + low_26) / 2.0

    return {"tenkan": round(tenkan, 5), "kijun": round(kijun, 5)}


def calculate_swing_points(highs, lows, window=2):
    """
    Detects Williams Fractal Swing Highs and Swing Lows.
    A Swing High at bar i requires: High[i] > High[i-k] and High[i] > High[i+k] for k in 1..window.
    A Swing Low at bar i requires: Low[i] < Low[i-k] and Low[i] < Low[i+k] for k in 1..window.
    """
    if not highs or not lows or len(highs) < (2 * window + 1):
        return {
            "swing_highs": [],
            "swing_lows": [],
            "last_swing_high": highs[-1] if highs else None,
            "last_swing_low": lows[-1] if lows else None,
        }

    swing_highs = []
    swing_lows = []

    for i in range(window, len(highs) - window):
        is_sh = True
        is_sl = True
        for w in range(1, window + 1):
            if highs[i] <= highs[i - w] or highs[i] <= highs[i + w]:
                is_sh = False
            if lows[i] >= lows[i - w] or lows[i] >= lows[i + w]:
                is_sl = False

        if is_sh:
            swing_highs.append({"price": highs[i], "idx": i})
        if is_sl:
            swing_lows.append({"price": lows[i], "idx": i})

    last_sh = swing_highs[-1]["price"] if swing_highs else max(highs[-10:])
    last_sl = swing_lows[-1]["price"] if swing_lows else min(lows[-10:])

    return {
        "swing_highs": swing_highs,
        "swing_lows": swing_lows,
        "last_swing_high": last_sh,
        "last_swing_low": last_sl,
    }


def calculate_vsa_metrics(highs, lows, closes, volumes=None):
    """
    Calculates Volume Spread Analysis (VSA) metrics.
    If real tick volumes are available, uses real volumes; otherwise uses ATR-normalized candle range.
    Returns dict with spread, relative volume, effort vs result, and accumulation/distribution flags.
    """
    if not closes or len(closes) < 10:
        return {
            "relative_volume": 1.0,
            "relative_spread": 1.0,
            "is_ultra_high_vol": False,
            "is_narrow_spread": False,
            "effort_result_divergence": False,
            "vsa_bias": "NEUTRAL",
        }

    ranges = [highs[i] - lows[i] for i in range(len(closes))]
    curr_range = ranges[-1]
    avg_range = sum(ranges[-10:]) / 10.0 if sum(ranges[-10:]) > 0 else 1e-5
    relative_spread = curr_range / avg_range

    if volumes and len(volumes) >= 10 and sum(volumes[-10:]) > 0:
        curr_vol = volumes[-1]
        avg_vol = sum(volumes[-10:]) / 10.0
    else:
        # ATR-normalized proxy volume
        curr_vol = curr_range * 10000.0
        avg_vol = avg_range * 10000.0

    avg_vol = max(1e-5, avg_vol)
    relative_volume = curr_vol / avg_vol

    is_ultra_high_vol = relative_volume >= 1.4
    is_narrow_spread = relative_spread <= 0.65
    effort_result_divergence = is_ultra_high_vol and is_narrow_spread

    vsa_bias = "NEUTRAL"
    if effort_result_divergence:
        # High volume but price didn't move -> absorption
        if closes[-1] > closes[-2]:
            vsa_bias = "ACCUMULATION"  # Buying effort absorbed at resistance / support
        else:
            vsa_bias = "DISTRIBUTION"

    return {
        "relative_volume": round(relative_volume, 2),
        "relative_spread": round(relative_spread, 2),
        "is_ultra_high_vol": is_ultra_high_vol,
        "is_narrow_spread": is_narrow_spread,
        "effort_result_divergence": effort_result_divergence,
        "vsa_bias": vsa_bias,
    }


def classify_market_regime(highs, lows, closes, period=20):
    """
    Statistically classifies the current market regime using ADX trend strength,
    EMA separation, and Bollinger Squeeze ratios.
    Returns a dict with complete regime analytics.
    """
    if not closes or len(closes) < 30:
        return {
            "regime": "RANGING",
            "detailed_regime": "RANGING_COMPRESSED",
            "volatility": "LOW",
            "trend_intensity": 0.0,
            "squeeze_ratio": 0.0,
            "adx": 20.0,
            "direction": "NEUTRAL",
        }

    ema_long_period = min(200, len(closes) - 1)
    ema_short_period = min(20, len(closes) - 1)

    ema_long = calculate_ema(closes, ema_long_period) or closes[-1]
    ema_short = calculate_ema(closes, ema_short_period) or closes[-1]
    atr_val = calculate_atr(highs, lows, closes, 14) or (closes[-1] * 0.001)

    adx_val = calculate_adx(highs, lows, closes, 14)

    # Trend Intensity = Abs distance between Short and Long EMA normalized by ATR
    trend_intensity = abs(ema_short - ema_long) / atr_val if atr_val > 0 else 0.0

    # Multi-factor trend decision: ADX > 22 AND (trend_intensity > 1.0 or EMA separation)
    is_trending = (adx_val >= 22.0 and trend_intensity >= 0.8) or (trend_intensity >= 1.5)
    regime = "TRENDING" if is_trending else "RANGING"

    direction = "BULLISH" if ema_short >= ema_long else "BEARISH"

    squeeze = calculate_bollinger_squeeze(closes, period, 2.0) or 0.0

    # Historical squeeze benchmark
    historical_squeezes = []
    lookback = min(40, len(closes))
    for i in range(len(closes) - lookback, len(closes)):
        sq = calculate_bollinger_squeeze(closes[:i], period, 2.0)
        if sq is not None:
            historical_squeezes.append(sq)

    avg_squeeze = (
        sum(historical_squeezes) / float(len(historical_squeezes))
        if historical_squeezes
        else squeeze
    )
    volatility = "HIGH" if squeeze > avg_squeeze else "LOW"

    if regime == "TRENDING":
        detailed_regime = f"TRENDING_{direction}"
    else:
        detailed_regime = "RANGING_EXPANDED" if volatility == "HIGH" else "RANGING_COMPRESSED"

    return {
        "regime": regime,
        "detailed_regime": detailed_regime,
        "volatility": volatility,
        "trend_intensity": round(trend_intensity, 4),
        "squeeze_ratio": round(squeeze, 4),
        "adx": round(adx_val, 2),
        "direction": direction,
    }


def get_smc_analysis(history_bars):
    """Returns institutional Smart Money Concepts (SMC) & ICT market structure analysis."""
    try:
        import institutional_integrations.smc_ict_engine as smc
        return smc.global_smc_engine.analyze(history_bars)
    except Exception:
        return {
            "order_blocks": {"bullish_ob": None, "bearish_ob": None},
            "fvgs": {"bullish_fvgs": [], "bearish_fvgs": []},
            "mss": {"mss_status": "NEUTRAL", "break_level": None},
            "liquidity_sweeps": {"bsl_sweep": False, "ssl_sweep": False},
            "bias": "NEUTRAL",
            "confluence_score": 50.0,
        }


def calculate_order_flow_metrics(history_bars, order_book=None):
    """
    Computes microstructure order flow signals combining VPIN flow toxicity,
    DOM level 2 depth imbalances, and short-term book pressure.
    """
    try:
        import institutional_integrations.order_flow_imbalance as ofi

        # Build proxy buy/sell volume buckets from price action and tick volumes
        vol_buys = []
        vol_sells = []
        for bar in history_bars[-20:]:
            high = bar.get("high", 0.0)
            low = bar.get("low", 0.0)
            close = bar.get("close", 0.0)
            open_p = bar.get("open", close)
            vol = bar.get("vol", bar.get("volume", (high - low) * 10000.0))

            rng = max(1e-5, high - low)
            body_ratio = (close - open_p) / rng if rng > 0 else 0.0

            # Estimate buy vs sell volume based on close relative to bar range
            buy_vol = vol * max(0.05, min(0.95, (close - low) / rng))
            sell_vol = vol - buy_vol
            vol_buys.append(buy_vol)
            vol_sells.append(sell_vol)

        avg_bucket = sum(vol_buys + vol_sells) / max(1.0, len(vol_buys) * 2.0)
        vpin = ofi.calculate_vpin(vol_buys, vol_sells, bucket_size=max(1.0, avg_bucket))

        dom_metrics = {"imbalance_ratio": 0.0, "dominant_side": "NEUTRAL"}
        pressure_metrics = {"pressure_score": 0.0, "expected_direction": "BALANCED"}

        if order_book and isinstance(order_book, dict):
            dom_metrics = ofi.detect_bid_ask_imbalance(order_book)
            pressure_metrics = ofi.predict_short_term_book_pressure(
                dom_metrics.get("total_bid_qty", 0.0),
                dom_metrics.get("total_ask_qty", 0.0),
            )

        return {
            "vpin": vpin,
            "is_toxic_flow": vpin >= 0.65,
            "dom_imbalance": dom_metrics.get("imbalance_ratio", 0.0),
            "dominant_side": dom_metrics.get("dominant_side", "NEUTRAL"),
            "pressure_score": pressure_metrics.get("pressure_score", 0.0),
            "expected_direction": pressure_metrics.get("expected_direction", "BALANCED"),
        }
    except Exception:
        return {
            "vpin": 0.15,
            "is_toxic_flow": False,
            "dom_imbalance": 0.0,
            "dominant_side": "NEUTRAL",
            "pressure_score": 0.0,
            "expected_direction": "BALANCED",
        }

File: modules/adapted/rust-finance/indicators.rs
// crates/signals/src/indicators.rs
// Zero-allocation technical indicators for the hot path.
// All functions operate on price slices and return computed values.

/// Exponential Moving Average (EMA)
/// smoothing = 2 / (period + 1)
pub fn ema(prices: &[f64], period: usize) -> Vec<f64> {
    if prices.is_empty() || period == 0 {
        return Vec::new();
    }
    let k = 2.0 / (period as f64 + 1.0);
    let mut result = Vec::with_capacity(prices.len());

    // Seed with SMA of first `period` values
    if prices.len() < period {
        return vec![prices.iter().sum::<f64>() / prices.len() as f64];
    }
    let sma: f64 = prices[..period].iter().sum::<f64>() / period as f64;
    for _ in 0..period - 1 {
        result.push(f64::NAN);
    }
    result.push(sma);

    for i in period..prices.len() {
        let prev = result[i - 1];
        let ema_val = prices[i] * k + prev * (1.0 - k);
        result.push(ema_val);
    }
    result
}

/// Simple Moving Average (SMA)
pub fn sma(prices: &[f64], period: usize) -> Vec<f64> {
    if prices.len() < period || period == 0 {
        return Vec::new();
    }
    let mut result = Vec::with_capacity(prices.len() - period + 1);
    let mut sum: f64 = prices[..period].iter().sum();
    result.push(sum / period as f64);
    for i in period..prices.len() {
        sum += prices[i] - prices[i - period];
        result.push(sum / period as f64);
    }
    result
}

/// Relative Strength Index (RSI)
/// Standard 14-period RSI with Wilder's smoothing
pub fn rsi(prices: &[f64], period: usize) -> Vec<f64> {
    if prices.len() < period + 1 || period == 0 {
        return Vec::new();
    }
    let mut gains = Vec::with_capacity(prices.len() - 1);
    let mut losses = Vec::with_capacity(prices.len() - 1);

    for i in 1..prices.len() {
        let change = prices[i] - prices[i - 1];
        if change > 0.0 {
            gains.push(change);
            losses.push(0.0);
        } else {
            gains.push(0.0);
            losses.push(-change);
        }
    }

    // Initial average gain/loss
    let mut avg_gain: f64 = gains[..period].iter().sum::<f64>() / period as f64;
    let mut avg_loss: f64 = losses[..period].iter().sum::<f64>() / period as f64;

    let mut result = Vec::with_capacity(prices.len() - period);

    // First RSI value
    let rs = if avg_loss == 0.0 {
        100.0
    } else {
        avg_gain / avg_loss
    };
    result.push(100.0 - (100.0 / (1.0 + rs)));

    // Subsequent values with Wilder's smoothing
    for i in period..gains.len() {
        avg_gain = (avg_gain * (period as f64 - 1.0) + gains[i]) / period as f64;
        avg_loss = (avg_loss * (period as f64 - 1.0) + losses[i]) / period as f64;
        let rs = if avg_loss == 0.0 {
            100.0
        } else {
            avg_gain / avg_loss
        };
        result.push(100.0 - (100.0 / (1.0 + rs)));
    }
    result
}

/// MACD (Moving Average Convergence Divergence)
/// Returns (macd_line, signal_line, histogram)
/// Default periods: fast=12, slow=26, signal=9
pub struct MacdResult {
    pub macd_line: Vec<f64>,
    pub signal_line: Vec<f64>,
    pub histogram: Vec<f64>,
}

pub fn macd(prices: &[f64], fast: usize, slow: usize, signal_period: usize) -> Option<MacdResult> {
    if prices.len() < slow {
        return None;
    }
    let ema_fast = ema(prices, fast);
    let ema_slow = ema(prices, slow);

    // MACD line = fast EMA - slow EMA (aligned from slow-1 index)
    let mut macd_line = Vec::with_capacity(prices.len());
    for i in 0..prices.len() {
        if i < slow - 1 || ema_fast[i].is_nan() || ema_slow[i].is_nan() {
            macd_line.push(f64::NAN);
        } else {
            macd_line.push(ema_fast[i] - ema_slow[i]);
        }
    }

    // Signal line = EMA of MACD line (skip NaN values)
    let valid_macd: Vec<f64> = macd_line.iter().filter(|v| !v.is_nan()).copied().collect();
    let signal_ema = ema(&valid_macd, signal_period);

    let mut signal_line = Vec::with_capacity(prices.len());
    let offset = prices.len() - valid_macd.len();
    for _ in 0..offset {
        signal_line.push(f64::NAN);
    }
    for i in 0..valid_macd.len() {
        if i < signal_ema.len() {
            signal_line.push(signal_ema[i]);
        } else {
            signal_line.push(f64::NAN);
        }
    }

    // Histogram = MACD - Signal
    let mut histogram = Vec::with_capacity(prices.len());
    for i in 0..prices.len() {
        if macd_line[i].is_nan() || i >= signal_line.len() || signal_line[i].is_nan() {
            histogram.push(f64::NAN);
        } else {
            histogram.push(macd_line[i] - signal_line[i]);
        }
    }

    Some(MacdResult {
        macd_line,
        signal_line,
        histogram,
    })
}

/// Bollinger Bands
/// Returns (upper, middle, lower)
pub struct BollingerBands {
    pub upper: Vec<f64>,
    pub middle: Vec<f64>,
    pub lower: Vec<f64>,
}

pub fn bollinger_bands(prices: &[f64], period: usize, num_std: f64) -> Option<BollingerBands> {
    if prices.len() < period {
        return None;
    }
    let middle = sma(prices, period);
    let mut upper = Vec::with_capacity(middle.len());
    let mut lower = Vec::with_capacity(middle.len());

    for (i, &mid) in middle.iter().enumerate() {
        let window = &prices[i..i + period];
        let variance: f64 = window.iter().map(|p| (p - mid).powi(2)).sum::<f64>() / period as f64;
        let std_dev = variance.sqrt();
        upper.push(mid + num_std * std_dev);
        lower.push(mid - num_std * std_dev);
    }

    Some(BollingerBands {
        upper,
        middle,
        lower,
    })
}

/// Volume Weighted Average Price (VWAP)
/// prices and volumes must be same length
pub fn vwap(prices: &[f64], volumes: &[f64]) -> Vec<f64> {
    if prices.len() != volumes.len() || prices.is_empty() {
        return Vec::new();
    }
    let mut cumulative_pv = 0.0;
    let mut cumulative_vol = 0.0;
    let mut result = Vec::with_capacity(prices.len());

    for i in 0..prices.len() {
        cumulative_pv += prices[i] * volumes[i];
        cumulative_vol += volumes[i];
        if cumulative_vol > 0.0 {
            result.push(cumulative_pv / cumulative_vol);
        } else {
            result.push(0.0);
        }
    }
    result
}

/// Average True Range (ATR)
/// Requires high, low, close arrays of equal length
pub fn atr(highs: &[f64], lows: &[f64], closes: &[f64], period: usize) -> Vec<f64> {
    if highs.len() != lows.len() || lows.len() != closes.len() || closes.len() < period + 1 {
        return Vec::new();
    }

    let mut true_ranges = Vec::with_capacity(closes.len() - 1);
    for i in 1..closes.len() {
        let hl = highs[i] - lows[i];
        let hc = (highs[i] - closes[i - 1]).abs();
        let lc = (lows[i] - closes[i - 1]).abs();
        true_ranges.push(hl.max(hc).max(lc));
    }

    // First ATR = SMA of first `period` true ranges
    let mut result = Vec::with_capacity(true_ranges.len());
    let first_atr: f64 = true_ranges[..period].iter().sum::<f64>() / period as f64;
    for _ in 0..period - 1 {
        result.push(f64::NAN);
    }
    result.push(first_atr);

    // Wilder's smoothing
    for i in period..true_ranges.len() {
        let prev = result[i - 1];
        let atr_val = (prev * (period as f64 - 1.0) + true_ranges[i]) / period as f64;
        result.push(atr_val);
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sma_basic() {
        let prices = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        let result = sma(&prices, 3);
        assert_eq!(result.len(), 3);
        assert!((result[0] - 2.0).abs() < 1e-10);
        assert!((result[1] - 3.0).abs() < 1e-10);
        assert!((result[2] - 4.0).abs() < 1e-10);
    }

    #[test]
    fn test_ema_basic() {
        let prices = vec![10.0, 11.0, 12.0, 13.0, 14.0, 15.0];
        let result = ema(&prices, 3);
        assert_eq!(result.len(), prices.len());
        assert!(result[2].is_finite());
    }

    #[test]
    fn test_rsi_basic() {
        // Monotonically increasing prices should give RSI near 100
        let prices: Vec<f64> = (0..30).map(|i| 100.0 + i as f64).collect();
        let result = rsi(&prices, 14);
        assert!(!result.is_empty());
        assert!(result.last().unwrap() > &90.0);
    }

    #[test]
    fn test_rsi_decreasing() {
        // Monotonically decreasing prices should give RSI near 0
        let prices: Vec<f64> = (0..30).map(|i| 200.0 - i as f64).collect();
        let result = rsi(&prices, 14);
        assert!(!result.is_empty());
        assert!(result.last().unwrap() < &10.0);
    }

    #[test]
    fn test_bollinger_bands() {
        let prices: Vec<f64> = (0..20).map(|i| 100.0 + (i as f64).sin() * 5.0).collect();
        let bands = bollinger_bands(&prices, 10, 2.0).unwrap();
        assert_eq!(bands.upper.len(), bands.middle.len());
        assert_eq!(bands.middle.len(), bands.lower.len());
        for i in 0..bands.upper.len() {
            assert!(bands.upper[i] >= bands.middle[i]);
            assert!(bands.middle[i] >= bands.lower[i]);
        }
    }

    #[test]
    fn test_vwap() {
        let prices = vec![100.0, 101.0, 102.0, 103.0];
        let volumes = vec![1000.0, 2000.0, 1500.0, 500.0];
        let result = vwap(&prices, &volumes);
        assert_eq!(result.len(), 4);
        // First VWAP = 100.0
        assert!((result[0] - 100.0).abs() < 1e-10);
    }

    #[test]
    fn test_macd() {
        let prices: Vec<f64> = (0..50)
            .map(|i| 100.0 + (i as f64 * 0.5).sin() * 10.0)
            .collect();
        let result = macd(&prices, 12, 26, 9);
        assert!(result.is_some());
        let r = result.unwrap();
        assert_eq!(r.macd_line.len(), prices.len());
    }
}
